auth_saml/controllers/main.py

```diff
@@ -0,0 +1,225 @@
+                request.cr, SUPERUSER_ID, [('enabled', '=', True)]
+            )
+        except Exception, e:
+            _logger.exception("SAML2: %s" % str(e))
+            providers = []
+
+        return providers
+
+    @http.route()
+    def web_login(self, *args, **kw):
+        ensure_db()
+        if (
+            request.httprequest.method == 'GET' and
+            request.session.uid and
+            request.params.get('redirect')
+        ):
+
+            # Redirect if already logged in and redirect param is present
+            return http.redirect_with_hash(request.params.get('redirect'))
+
+        providers = self.list_providers()
+
+        response = super(SAMLLogin, self).web_login(*args, **kw)
+        if response.is_qweb:
+            error = request.params.get('saml_error')
+            if error == '1':
+                error = _("Sign up is not allowed on this database.")
+            elif error == '2':
+                error = _("Access Denied")
+            elif error == '3':
+                error = _(
+                    "You do not have access to this database or your "
+                    "invitation has expired. Please ask for an invitation "
+                    "and be sure to follow the link in your invitation email."
+                )
+            else:
+                error = None
+
+            response.qcontext['providers'] = providers
+
+            if error:
+                response.qcontext['error'] = error
+
+        return response
+
+
+class AuthSAMLController(http.Controller):
+
+    def get_state(self, provider_id):
+        redirect = request.params.get('redirect') or 'web'
+        if not redirect.startswith(('//', 'http://', 'https://')):
+            redirect = '%s%s' % (
+                request.httprequest.url_root,
+                redirect[1:] if redirect[0] == '/' else redirect
+            )
+
+        state = {
+            "d": request.session.db,
+            "p": provider_id,
+            "r": werkzeug.url_quote_plus(redirect),
+        }
+        return state
+
+    @http.route('/auth_saml/get_auth_request', type='http', auth='none')
+    def get_auth_request(self, pid):
+        """state is the JSONified state object and we need to pass
+        it inside our request as the RelayState argument
+        """
+
+        provider_id = int(pid)
+        provider_osv = request.registry.get('auth.saml.provider')
+
+        auth_request = None
+        state = self.get_state(provider_id)
+
+        try:
+            auth_request = provider_osv._get_auth_request(
+                request.cr, SUPERUSER_ID, provider_id, state
+            )
+
+        except Exception, e:
+            _logger.exception("SAML2: %s" % str(e))
+
+        # TODO: handle case when auth_request comes back as None
+
+        # store a RelayState on the request to our IDP so that the IDP
+        # can send us back this info alongside the obtained token
+        params = {
```

Comment:
If there is already a RelayState parameter, this method adds another so the URL will contain 2 RelayState. Since SAML use this parameter in order to validate the Signature, the IdP will experience issues on its signature validation.

A fast patch for that, is simply to add the RelayState if no RelayState already present:
```
        url = auth_request
        if not ('RelayState' in auth_request):
            params = {
                "RelayState": simplejson.dumps({
                    "d": request.session.db,
                    "p": pid,
                }),
            }
            url += "&" + werkzeug.url_encode(params)

```